nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/pubsub/publish/FlowFileResult.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.nifi.processors.gcp.pubsub.publish;
+
+import com.google.api.core.ApiFuture;
+import com.google.api.core.ApiFutures;
+import com.google.api.gax.rpc.UnavailableException;
+import org.apache.nifi.flowfile.FlowFile;
+import org.apache.nifi.processor.Relationship;
+import org.apache.nifi.processors.gcp.pubsub.PubSubAttributes;
+import org.apache.nifi.processors.gcp.pubsub.PublishGCPubSub;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+
+/**
+ * Tracking of an interaction from NiFi {@link PublishGCPubSub} processor to Google PubSub endpoint.
+ */
+public class FlowFileResult {
+    private static final Logger logger = LoggerFactory.getLogger(FlowFileResult.class);
+
+    private final FlowFile flowFile;
+    private final Map<String, String> attributes;
+    private final List<ApiFuture<String>> futures;
+    private final List<String> successes;
+    private final List<Throwable> failures;
+
+    public FlowFileResult(final FlowFile flowFile, final List<ApiFuture<String>> futures,
+                          final List<String> successes, final List<Throwable> failures) {
+        this.flowFile = flowFile;
+        this.attributes = new HashMap<>();
+        this.futures = futures;
+        this.successes = successes;
+        this.failures = failures;
+    }
+
+    /**
+     * After all in-flight messages have results, calculate appropriate {@link Relationship}.
+     */
+    public Relationship reconcile() {
+        while (futures.size() > (successes.size() + failures.size())) {
+            try {
+                ApiFutures.allAsList(futures).get();
+            } catch (InterruptedException | ExecutionException e) {
+                logger.error("Error while reconciling PublishGCPubSub send", e);
```

Comment:
Recommend adjusting the wording slightly:
```suggestion
                logger.error("Failed to reconcile PubSub send operation status", e);
```